File: cpp/topKFrequentElement.cpp
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

template<typename T>
void print_vector(vector<T> vec){
    for(T v : vec){
        cout << v << "\n";
    }
    cout << endl;
}
template<typename T,typename U>
void print_map(unordered_map<T,U> map){
    for(auto entry : map){
        cout << entry.first << " <===> " << entry.second << " \n";
    }
    cout << endl;
}

class StackSort{
    public:
        vector<int> sort_stack(vector<int> stk){
            if (stk.size()==0)
                return stk;
            int top = stk.back();
            stk.pop_back();
            sort_stack(stk);
            this->insert_in_sorted_stack(stk,top);
            stk.push_back(top);
            return stk;
        }

        void insert_in_sorted_stack(vector<int> stk,int value){
            if(stk.size()==0 || stk.back()<= value){
                stk.push_back(value);
                return;
            }
            int top = stk.back();
            stk.pop_back();
            insert_in_sorted_stack(stk,value);
            stk.push_back(top);
        }
};


class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> num_to_freq;
        unordered_map<int,int> freq_to_num;
        for(int num:nums){
            if(num_to_freq.find(num)==num_to_freq.end())
                num_to_freq[num] = 1;
            else
                num_to_freq[num]++;
        }

        vector<int> ret;
        for (int num:nums){
            freq_to_num[num_to_freq[num]] = num;
        }

        for(auto entry : freq_to_num){
            ret.push_back(entry.first);
        }
        StackSort stk;
        ret = stk.sort_stack(ret);
        vector<int> ans;
        int i;
        for(i=0;i<k;i++){
            ans.push_back(freq_to_num[ret.back()]);
            ret.pop_back();
        }
        return ans;
    }
}; 

File: python/topKFrequentElement.py


# 21 / 21 test cases passed.
# Status: Accepted
# Runtime: 882 ms
# Memory Usage: 19.7 MB

from typing import List

def sortStack(stack):
    if len(stack)==0:
        return stack
    top = stack.pop()
    sortStack(stack)
    insertInSortedOrder(stack,top)
    return stack

def insertInSortedOrder(stack,value):
    if len(stack) == 0 or stack[-1][1] <= value[1]:
        stack.append(value)
        return
    top=stack.pop()
    insertInSortedOrder(stack, value)
    stack.append(top)




class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        table = {}
        for num in nums:
            if num not in table:
                table[num] = 1 
            else:
                table[num]+=1 
        r = sortStack([[k,v] for k,v in table.items()])
        ans = []
        i = 0;
        while i < k:
            ans.append(r.pop()[0])
            i += 1
        return ans



